### ISSUE TITLE
[Discussion] Expected behavior of logical expressions in Taichi

### ISSUE THREAD

**Issue by strongoier** [taichi-dev/taichi#3635]
Currently, there isn't a clear description of expected behavior of logical expressions (`and` / `or`) in Taichi. The implementation, https://github.com/taichi-dev/taichi/blob/71dd17426b62327864a21c65ca570812997c33ba/python/taichi/lang/ops.py#L792-L793, shows that `logical_and` / `logical_or` is simply treated as `bitwise_and` / `bitwise_or`. This leads to the non-short-circuit behavior (#3572) and the non-intuitive computation result:
```python
@ti.kernel
def func(a: ti.i32, b: ti.i32) -> ti.i32:
    return a and b

print(func(2, 3))  # 2
```

In Python, the behavior of logical expressions is like:
```python
print(2 and 3)  # 3
print(2.0 or 0)  # 2.0
```
which is short-circuiting and uses the `True` expression as the result.

Should we make logical expressions in Taichi behave the same as in Python, keep the original behavior in Taichi, or define yet-another behavior? Opinions are welcome.

**Comment by lin-hitonami** [comment]
I think the behavior should be identical to Python's because most Python users would expect taichi's grammar is similar to Python's.

**Comment by ailzhang** [comment]
I second @lin-hitonami 's idea, python frontend's behavior should be close to python convention as much as possible. ;)

**Comment by strongoier** [comment]
According to the discussions in this line, we will make the behavior of logical expressions in Taichi the same as Python.

**Comment by lin-hitonami** [comment]
There is a problem if we make the behavior the same as Python. For this program:
```python
@ti.kernel
def foo(a: ti.i32, b: ti.f32):
    c = a and b
```
What is the type of `c`? Will this confuse the type checker?

**Comment by strongoier** [comment]
Yes this is indeed a problem. We cannot have exactly the same behavior as Python because Taichi is statically typed. In Taichi, all operands of logical expressions should be enforced to have a same type.

Then the core problem is more related to the order of our compilation process - we should do type check first, and then do transformations which may lose type enforcement information.
